Lib/concurrent/futures/process.py

```diff
@@ -256,122 +258,102 @@ def _process_worker(call_queue, result_queue, initializer, initargs):
-    """Manages the communication between this process and the worker processes.
+        # When the executor gets garbarge collected, the weakref callback
+        # will wake up the queue management thread so that it can terminate
+        # if there is no pending work item.
+        self.thread_wakeup = executor._executor_manager_thread_wakeup
 
-    This function is run in a local thread.
+        def weakref_cb(_, thread_wakeup=self.thread_wakeup):
+            mp.util.debug('Executor collected: triggering callback for'
+                          ' QueueManager wakeup')
+            thread_wakeup.wakeup()
 
-    Args:
-        executor_reference: A weakref.ref to the ProcessPoolExecutor that owns
-            this thread. Used to determine if the ProcessPoolExecutor has been
-            garbage collected and that this function can exit.
-        process: A list of the ctx.Process instances used as
-            workers.
-        pending_work_items: A dict mapping work ids to _WorkItems e.g.
-            {5: <_WorkItem...>, 6: <_WorkItem...>, ...}
-        work_ids_queue: A queue.Queue of work ids e.g. Queue([5, 6, ...]).
-        call_queue: A ctx.Queue that will be filled with _CallItems
-            derived from _WorkItems for processing by the process workers.
-        result_queue: A ctx.SimpleQueue of _ResultItems generated by the
-            process workers.
-        thread_wakeup: A _ThreadWakeup to allow waking up the
-            queue_manager_thread from the main Thread and avoid deadlocks
-            caused by permanently locked queues.
-    """
-    executor = None
+        self.executor_reference = weakref.ref(executor, weakref_cb)
 
-    def shutting_down():
-        return (_global_shutdown or executor is None
-                or executor._shutdown_thread)
+        # Store references to necessary internals of the executor.
+        self.processes = executor._processes
+        self.call_queue = executor._call_queue
+        self.result_queue = executor._result_queue
+        self.work_ids_queue = executor._work_ids
+        self.pending_work_items = executor._pending_work_items
 
-    def shutdown_worker():
-        # This is an upper bound on the number of children alive.
-        n_children_alive = sum(p.is_alive() for p in processes.values())
-        n_children_to_stop = n_children_alive
-        n_sentinels_sent = 0
-        # Send the right number of sentinels, to make sure all children are
-        # properly terminated.
-        while n_sentinels_sent < n_children_to_stop and n_children_alive > 0:
-            for i in range(n_children_to_stop - n_sentinels_sent):
-                try:
-                    call_queue.put_nowait(None)
-                    n_sentinels_sent += 1
-                except Full:
-                    break
-            n_children_alive = sum(p.is_alive() for p in processes.values())
+        # Set this thread to be daemonized
+        super().__init__()
+        self.daemon = True
```

Comment:
I think it would be worthwhile exploring whether/how we can remove the daemon flag, though in another BPO issue.

Reply:
From my understanding, the daemon flag is only necessary if we can't guarantee that the thread will be joined at the end of the shutdown process. But, I could see this potentially being an issue when calling `executor.shutdown(wait=False)` for ProcessPoolExecutor (which has issues with graceful termination).

Reply:
Actually, I am not sure this is needed anymore.
The `shutdown(wait=False)` behavior should be fixed by #17670 and the clean up process is quite robust I believe. We should maybe give a try to simply removing the flag and experiments bad situations to see if it leads to interpreters hanging at exit.

Reply:
The problem is: what happens if the interpreter starts to shutdown while an executor is still running? If the management thread is not daemon, the interpreter will hang trying to join it.

Reply:
I think this is the current behavior. The `_python_exit` function will join the thread anyway no?

Reply:
The problem is that `atexit` hooks are called after non-daemon threads are joined. See https://bugs.python.org/issue37266#msg362890

Reply:
Ah sorry, I missed these messages.
Indeed, the simplest solution would be to call the atexit `_python_exit` before joining non-deamonized threads.